src/main/kotlin/org/rust/ide/completion/AttributeCompletionProvider.kt

```diff
@@ -0,0 +1,164 @@
+        RustAttribute("should_panic", onTestFn),
+        RustAttribute("cold", onFn),
+        RustAttribute("naked", onFn),
+        RustAttribute("thread_local", onStaticMut),
+        RustAttribute("link_args", onExternBlock), // Feature gated as of 1.9
+        RustAttribute("link", onExternBlock),
+        RustAttribute("linked_from", onExternBlock), // Feature gated as of 1.9
+        RustAttribute("link_name", onExternBlockDecl),
+        RustAttribute("linkage", onExternBlockDecl),
+        RustAttribute("repr", onStruct or onEnum),
+        RustAttribute("macro_use", onMod or onExternCrate),
+        RustAttribute("macro_reexport", onExternCrate),
+        RustAttribute("macro_export", onMacro),
+        RustAttribute("no_link", onExternCrate),
+        RustAttribute("export_name", onStatic or onFn),
+        RustAttribute("link_section", onStatic or onFn),
+        RustAttribute("no_mangle", onAnyItem),
+        RustAttribute("simd", onTupleStruct), // Feature gated as of 1.9
+        RustAttribute("unsafe_destructor_blind_to_params", onDropFn), // Feature gated as of 1.9
+        RustAttribute("unsafe_no_drop_flag", onStruct), // Feature gated as of 1.9
+        RustAttribute("doc", onAnyItem),
+        RustAttribute("rustc_on_unimplemented", onTrait), // Feature gated as of 1.9
+        RustAttribute("cfg", onFn),
+        RustAttribute("cfg_attr", onAnyItem),
+        RustAttribute("allow", onAnyItem),
+        RustAttribute("warn", onAnyItem),
+        RustAttribute("forbid", onAnyItem),
+        RustAttribute("deny", onAnyItem),
+        RustAttribute("lang", onFn),
+        RustAttribute("inline", onFn),
+        RustAttribute("derive", onStruct or onEnum))
+
+    override fun addCompletions(parameters: CompletionParameters,
+                                context: ProcessingContext?,
+                                result: CompletionResultSet) {
+
+        val suggestions = attributes.filter { it.appliesTo.accepts(parameters.position)}
+            .map { LookupElementBuilder.create(it.name) }
+        result.addAllElements(suggestions)
+    }
+
+    val elementPattern: ElementPattern<PsiElement> get() {
+        val outerAttrElem = psiElement(RustCompositeElementTypes.OUTER_ATTR)
+        val innerAttrElem = psiElement(RustCompositeElementTypes.INNER_ATTR)
+        val metaItemElem = psiElement(RustCompositeElementTypes.META_ITEM)
+            .andOr(psiElement().withParent(outerAttrElem), psiElement().withParent(innerAttrElem))
+        return psiElement().withParent(metaItemElem).withLanguage(RustLanguage);
+    }
+
+    private data class RustAttribute(val name: String, val appliesTo: ElementPattern<PsiElement>)
+
+    private val onStruct: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustStructItemElement::class.java)
+    }
+
+    private val onEnum: ElementPattern<PsiElement> get() {
+        return psiElement().withSuperParent(3, RustEnumItemElement::class.java)
```

Comment:
Good call.  I'm not sure what you're suggesting with the inline function though, which may be due the fact that I'm relatively new to Kotlin.

Would a standard function accomplish the same thing?

    private fun onItem(aClass: Class<out PsiElement>): ElementPattern<PsiElement> {
        return psiElement().withSuperParent(3, aClass)
    }